pkg/operator/operator.go

```diff
@@ -241,76 +273,213 @@ func (k *Kontroller) withLeaderElection() error {
 func (k *Kontroller) process() {
 	glog.V(4).Info("Going through a loop cycle")
 
-	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	// find nodes with the after-reboot=true label and check if all provided
+	// annotations are set. if all annotations are set to true then remove the
+	// after-reboot=true label and set reboot-ok=false, telling the agent that
+	// the reboot has completed.
+	err := k.checkAfterReboot()
 	if err != nil {
-		glog.Infof("Failed listing nodes %v", err)
+		glog.Errorf("Failed to check after reboot: %v", err)
 		return
 	}
 
-	glog.V(6).Infof("Found nodes: %+v", nodelist.Items)
+	// find nodes which just rebooted but haven't run after-reboot checks.
+	// remove any old annotations and add the after-reboot=true label.
+	err = k.markAfterReboot()
+	if err != nil {
+		glog.Errorf("Failed to update recently rebooted nodes: %v", err)
+		return
+	}
 
-	justRebootedNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, justRebootedSelector)
+	// find nodes with the before-reboot=true label and check if all provided
+	// annotations are set. if all annotations are set to true then remove the
+	// before-reboot=true label and set reboot=ok=true, telling the agent it's
+	// time to reboot.
+	err = k.checkBeforeReboot()
```

Comment:
Discussed. Its clear that these 4 state transition functions can run in different orders and the current ordering is coincidental and errors are divided btw serious errors worth waiting for the process loop and those that are more benign. 

I'd favor these methods returning any error that occurs though. Its easier to explain and to write tests if the errors are returned rather than being log lines. Waiting for the process loop shouldn't be too common since the places we can error now already have reties baked in and if we switch to Watches in the future, the fixed period goes away anyway. If we instead plan for them to be 4 independent loops, it'd be ok to spawn 4 goroutines for them, that way nobody starts depending on their ordering.